pkg/virt-controller/watch/vmi.go

```diff
@@ -1284,40 +1299,108 @@ func (c *VMIController) handleHotplugVolumes(hotplugVolumes []*virtv1.Volume, ho
 	if len(newVolumes) == 0 {
 		return nil
 	}
+	if len(newVolumes) > 0 {
+		// New volumes detected, create hotplug pods.
+		for _, volume := range newVolumes {
+			logger.V(1).Infof("Processing new hotplugged volume: %s", volume.Name)
+			var attachmentPodTemplate *k8sv1.Pod
+			var err error
+			ready, wffc, err := c.volumeReadyToUse(vmi.Namespace, *volume, dataVolumes)
+			if err != nil {
+				return &syncErrorImpl{fmt.Errorf("Error determining volume status %v", err), PVCNotReadyReason}
+			}
+			if wffc {
+				// Volume in WaitForFirstConsumer, it has not been populated by CDI yet. create a dummy pod
+				logger.V(3).Infof("Volume %s/%s is in WaitForFistConsumer, triggering population", vmi.Namespace, volume.Name)
+				syncError := c.triggerHotplugPopulation(volume, vmi, virtLauncherPod)
+				if syncError != nil {
+					return syncError
+				}
+				continue
+			}
+			if !ready {
+				// Volume not ready, skip until it is.
+				logger.V(3).Infof("Skipping hotplugged volume: %s, not ready", volume.Name)
+				continue
+			}
+			// Check if the VMI VolumeStatus contains this volume, if that is the case then something deleted the attachment pod
+			// and we need to stop the VMI as that is a critical error.
+			if c.volumeStatusContainsVolumeAndPod(vmi.Status.VolumeStatus, volume) {
+				logger.V(1).Infof("Detected attachment pod is missing for VMI %s/%s, the VMI will be deleted", vmi.Namespace, vmi.Name)
+				return &syncErrorImpl{fmt.Errorf("Missing pod for hotplugged volume %s", volume.Name), MissingAttachmentPodReason}
+			}
+			attachmentPodTemplate, err = c.createAttachmentPodTemplate(volume, virtLauncherPod, vmi)
+			if attachmentPodTemplate == nil { // nil means the PVC is not populated yet.
+				continue
```

Comment:
it looks inconsistent here to have `triggerHotplugPopulation` both render and create via api the pod, while a few lines down we have a function `createAttachmentPodTemplate` that only renders the pod and the api is called directly in the top level function.

Reply:
Yeah should put that in a function as well. Done